quickwit/quickwit-common/src/sorted_iter.rs

```diff
@@ -0,0 +1,294 @@
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+use std::cmp::Ordering;
+use std::collections::{btree_map, btree_set};
+use std::iter::Peekable;
+
+/// Marks sorted iterators, typically iterators over [`BTreeSet`] and [`BTreeMap`].
+trait Sorted {}
+
+/// Defines helper methods on sorted iterators.
+pub trait SortedIterator: Iterator + Sized {
+    /// Compares two sorted iterators and returns the diff.
+    fn diff<U>(self, other: U) -> DiffIterator<Self, U>
+    where U: SortedIterator<Item = Self::Item> {
+        DiffIterator {
+            left: self.peekable(),
+            right: other.peekable(),
+        }
+    }
+}
+
+#[derive(Debug, Eq, PartialEq)]
+pub enum Diff<K> {
+    Added(K),
+    Unchanged(K),
+    Removed(K),
+}
+
+pub struct DiffIterator<T: Iterator, U: Iterator> {
+    left: Peekable<T>,
+    right: Peekable<U>,
+}
+
+impl<T, U, K> Iterator for DiffIterator<T, U>
+where
+    T: Iterator<Item = K>,
+    U: Iterator<Item = K>,
+    K: Ord,
+{
+    type Item = Diff<K>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        match (self.left.peek(), self.right.peek()) {
+            (Some(left), Some(right)) => match left.cmp(right) {
+                Ordering::Less => {
+                    let left = self
+                        .left
+                        .next()
+                        .expect("The left iterator should not be empty.");
+                    Some(Diff::Removed(left))
+                }
+                Ordering::Equal => {
+                    let left = self
+                        .left
+                        .next()
+                        .expect("The left iterator should not be empty.");
+                    self.right.next();
+                    Some(Diff::Unchanged(left))
+                }
+                Ordering::Greater => {
+                    let right = self
+                        .right
+                        .next()
+                        .expect("The right iterator should not be empty.");
+                    Some(Diff::Added(right))
+                }
+            },
+            (Some(_), None) => {
+                let left = self
+                    .left
+                    .next()
+                    .expect("The left iterator should not be empty.");
+                Some(Diff::Removed(left))
+            }
+            (None, Some(_)) => {
+                let right = self
+                    .right
+                    .next()
+                    .expect("The right iterator should not be empty.");
+                Some(Diff::Added(right))
+            }
+            (None, None) => None,
+        }
+    }
+}
+
+impl<T> SortedIterator for T where T: Iterator + Sorted {}
```

Comment:
probably a silly question, but why not just have a `SortedIterator` trait?
(and not have a Sorted trait).

Would implementating

```rust
impl<K, V> SortedIterator for btree_map::IntoKeys<K, V> {}
...
```
have worked?